src/opnsense/scripts/unbound/blocklists/__init__.py

```diff
@@ -0,0 +1,250 @@
+                        if parts[-1] != "\n":
+                            prev_chop = parts.pop()
+                        else:
+                            prev_chop = ''
+                        for part in parts:
+                            yield part
+                except Exception as e:
+                    syslog.syslog(syslog.LOG_ERR,'blocklist download : error reading file from %s (error : %s)' % (uri, e))
+                    return
+
+        else:
+            syslog.syslog(syslog.LOG_ERR,
+                'blocklist download : unable to download file from %s (status_code: %d)' % (uri, req.status_code)
+            )
+
+class BlocklistParser:
+    def __init__(self):
+        # check for a running download process, this may take a while so it's better to check...
+        try:
+            lck = open('/tmp/unbound-download_blocklists.tmp', 'w+')
+            fcntl.flock(lck, fcntl.LOCK_EX | fcntl.LOCK_NB)
+        except IOError:
+            # already running, exit status 99
+            sys.exit(99)
+
+        syslog.openlog('unbound', facility=syslog.LOG_LOCAL4)
+        self.handlers = list()
+        self._register_handlers()
+        self.startup_time = time.time()
+
+    def _register_handlers(self):
+        handlers = list()
+        for filename in glob.glob("%s/*.py" % os.path.dirname(__file__)):
+            importlib.import_module(".%s" % os.path.splitext(os.path.basename(filename))[0], __name__)
+
+        for module_name in dir(sys.modules[__name__]):
+            for attribute_name in dir(getattr(sys.modules[__name__], module_name)):
+                cls = getattr(getattr(sys.modules[__name__], module_name), attribute_name)
+                if isinstance(cls, type) and issubclass(cls, BaseBlocklistHandler)\
+                        and cls not in (BaseBlocklistHandler,):
+                    handlers.append(cls())
+        self.handlers = handlers
+
+    def _get_config(self):
+        for handler in self.handlers:
+            cfg = handler.get_config()
+            if cfg:
+                return cfg
+
+    def _merge_results(self, blocklists):
+        """
+        Take output of all the handlers and merge based on each handlers' priority.
+        The default handler has highest priority
+        """
+        if len(blocklists) == 1:
+            return next(iter(blocklists.values()))
+
+        blocklists = dict(sorted(blocklists.items(), reverse=True))
+        first = next(iter(blocklists.values()))
+        for bl in list(blocklists.values())[1:]:
+            for key, value in bl.items():
+                if key not in first:
+                    # no collision, merge
+                    first[key] = value
+                else:
+                    # a handler with a lower priority has provided a policy
+                    # on a domain that already exists in the blocklist,
+                    # add it for debugging purposes
+                    first[key].setdefault('collisions', []).append(value)
```

Comment:
I'm having some difficulties wrapping my head around this, how is priority handled? If for example the first handler offers a list with networks and the next doesn't, which one prevails?

Reply:
Priority is determined by a simple `priority` attribute in the plugged in class

Reply:
ah, got it